File: apps/api/app/db/models/task.py
import enum
from datetime import datetime, timezone
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Enum, ForeignKey, JSON, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.db.models.processor import Processor


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


class TaskStatus(str, enum.Enum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class ProcessingTask(Base):
    __tablename__ = "processing_tasks"

    id: Mapped[int] = mapped_column(primary_key=True)
    demand_id: Mapped[int] = mapped_column(ForeignKey("demands.id"), index=True)
    created_by: Mapped[int] = mapped_column(ForeignKey("users.id"), index=True)
    processor_id: Mapped[int | None] = mapped_column(
        ForeignKey("processors.id"),
        nullable=True,
        index=True,
    )
    task_type: Mapped[str] = mapped_column(String(32))
    status: Mapped[TaskStatus] = mapped_column(Enum(TaskStatus), default=TaskStatus.QUEUED)
    progress: Mapped[int] = mapped_column(default=0)
    config_json: Mapped[dict] = mapped_column(JSON)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
    )
    input_assets: Mapped[list["TaskInputAsset"]] = relationship(
        back_populates="task",
        cascade="all, delete-orphan",
        order_by="TaskInputAsset.id",
    )
    processor: Mapped["Processor | None"] = relationship()

    @property
    def input_asset_ids(self) -> list[int]:
        return [item.catalog_asset_id for item in self.input_assets]

    @property
    def processor_name(self) -> str | None:
        if self.processor is None:
            return None
        return self.processor.name


class TaskInputAsset(Base):
    __tablename__ = "task_input_assets"
    __table_args__ = (UniqueConstraint("task_id", "catalog_asset_id"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    task_id: Mapped[int] = mapped_column(ForeignKey("processing_tasks.id"), index=True)
    catalog_asset_id: Mapped[int] = mapped_column(ForeignKey("catalog_assets.id"), index=True)
    task: Mapped["ProcessingTask"] = relationship(back_populates="input_assets")


class TaskArtifact(Base):
    __tablename__ = "task_artifacts"

    id: Mapped[int] = mapped_column(primary_key=True)
    task_id: Mapped[int] = mapped_column(ForeignKey("processing_tasks.id"), index=True)
    artifact_type: Mapped[str] = mapped_column(String(64))
    file_name: Mapped[str] = mapped_column(String(255))
    file_path: Mapped[str] = mapped_column(String(255))
    sample_count: Mapped[int] = mapped_column()
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now)

File: apps/web/src/types/task.ts
export type TaskStatus = "queued" | "running" | "completed" | "failed";

export interface TaskItem {
  id: number;
  demandId: number;
  inputAssetIds: number[];
  createdBy: number;
  processorId?: number | null;
  processorName?: string | null;
  taskType: string;
  status: TaskStatus;
  progress: number;
  note: string | null;
  createdAt: string;
  config?: Record<string, string>;
}

export interface TaskCreatePayload {
  demandId: number;
  inputAssetIds: number[];
  taskType: string;
  config: Record<string, string>;
}

export interface TaskArtifactCreatePayload {
  artifactType: string;
  fileName: string;
  filePath: string;
  sampleCount: number;
  note?: string;
}

export interface TaskArtifactItem {
  id: number;
  taskId: number;
  artifactType: string;
  fileName: string;
  filePath: string;
  sampleCount: number;
  note: string | null;
  createdAt: string;
}

export const taskStatusLabels: Record<TaskStatus, string> = {
  queued: "排队中",
  running: "执行中",
  completed: "已完成",
  failed: "失败"
};

export const taskStatusTones: Record<TaskStatus, "warn" | "info" | "good" | "danger"> = {
  queued: "warn",
  running: "info",
  completed: "good",
  failed: "danger"
};

export const taskTypeLabels: Record<string, string> = {
  instruction: "指令生成",
  book_split: "拆书",
  grammar_fix: "病句修改"
};

export function formatTaskType(taskType: string): string {
  return taskTypeLabels[taskType] ?? taskType;
}
